src/main/native/GCM.c

```diff
@@ -1346,19 +1342,19 @@ JNIEXPORT jint JNICALL Java_com_ibm_crypto_plus_provider_ock_NativeInterface_do_
 	parm_block = parameters + ivLen + aadLen + tagLen + 8;
 	len = plaintextLen;
 	alen = aadLen;
-	
+
 	// Handle IV (different implementation based on the IV size)
 	handleIV(ivLen, keyLen, blockSize, J0Offset, (char*) iv, (char*) (parm_block + keyOffset), (char*) parm_block);
 	memcpy(parm_block + counterValueOffset, parm_block + J0Offset + blockSize - 4, 4); // Add Counter Value
 
-	zS390(input, output, aad, &len, &alen, parm_block, &mode);
+	zS390((input != NULL) ? (input + inputOffset) : NULL, (output != NULL) ? (output + outputOffset) : NULL, aad, &len, &alen, parm_block, &mode);
 
 	// Copy tag
 	tag = parameters + ivLen + aadLen + keyLen;
 	memcpy(tag, parm_block + 16, tagLen); // Add tag to output
 
-	if(inputJ != NULL) (*env)->ReleasePrimitiveArrayCritical(env, inputJ, input, 0);
-	if(outputJ) (*env)->ReleasePrimitiveArrayCritical(env, outputJ, output, 0);
+	if(inputJ != NULL) (*env)->ReleasePrimitiveArrayCritical(env, inputJ, (input != NULL) ? (input + inputOffset) : NULL, 0);
```

Comment:
Why does the `ReleasePrimitiveArrayCritical` have different parameters than the `GetPrimitiveArrayCritical`? Shouldn't we just release `input`?

Also, should we check `if(input != NULL)` instead of `if(inputJ != NULL)`?

Reply:
I mean that I think it should be 
`if(input != NULL) (*env)->ReleasePrimitiveArrayCritical(env, inputJ, input, 0);`
, since the get and release should be applied to the same arguments, not one with offset.